### PR TITLE
Introducing multi-staging area

### PR DESCRIPTION
**Pull request by seb-mtl**
## Description

This PR introduces a multi-staging area with a few additions:

- Index files are now stored in `.snow/indexes`
- Main index file is simply called `index`
- Custom index files are called `index.{hash}`
- Index files are now deleted upon commit
- Introduce `--index` for `snow add/log/commit/rm`
- New indexes can be created through `snow index create`

image

image

## Type of change

- [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

### CODE REVIEW

File: main.ts [hunk 1]
```diff
@@ -22,6 +23,23 @@ function fileMatch(relFilepath: string, relCwd: string, pathPattern: string): bo
...
 }
 
+/**
+ * Helper function to get an existing or creating a new index.
+ * @param index   Either null/defined, or 'create' or an existing index;
+*/
+function getIndex(repo: Repository, index: string | null | undefined) {
+  if (index === 'create') {
+    return repo.createIndex();
+  } if (index) {
```
Comment (daniellanner) [inline main.ts:33]: Add newline

File: src/index.ts [hunk 2]
```diff
@@ -26,49 +29,94 @@ export class Index {
...
+    }
+  }
+
+  /**
+   * Return the absolute path of the index directory
+   *
+   * @returns       Absolute path to the index directory.
+   */
+  static getAbsDir(repo: Repository): string {
+    const indexDir: string = join(repo.commondir(), 'indexes');
```
Comment (daniellanner) [inline src/index.ts:96]: opinion: prefer plural form 'indices'

File: src/repository.ts [hunk 3]
```diff
@@ -258,10 +261,62 @@ export class Repository {
...
+      this.repoIndexes.splice(foundIndex, 1);
+    }
+  }
+
+  /**
+   * Return all indexes. No copied are returned
    */
-  getIndex(): Index {
-    return this.repoIndex;
+  getIndexes(): Index[] {
```
Comment (daniellanner) [inline src/repository.ts:318]: opinion: prefer plural 'indices'

**Review by daniellanner — APPROVED** [review]
Add unit test for trying to re-stage files to multiple indices, should fail.

Otherwise, LGTM.

File: src/index.ts (hunk 2 above)
Comment (seb-mtl) [reply src/index.ts:96]: Agree, made me think about it but `Indexes` is more a term for databases https://www.essentialsql.com/what-is-a-database-index/

**Comment by seb-mtl** [comment]
> Add unit test for trying to re-stage files to multiple indices, should fail.
> 
> Otherwise, LGTM.

Added unit-tests: https://github.com/Snowtrack/SnowFS/pull/80/files#diff-6ccc5e2789c4c6912ae76adda84bbbbcd5f7b6e96e668a62252b1962ea0ab615R450

**Comment by daniellanner** [comment]
🎉

**Pull request merged**